ext/tidy/tidy.c

```diff
@@ -776,31 +714,74 @@ static void php_tidy_create_node(INTERNAL_FUNCTION_PARAMETERS, tidy_base_nodetyp
 	tidy_create_node_object(return_value, obj->ptdoc, node);
 }
 
+
+static bool php_tidy_set_tidy_opt(TidyDoc doc, const char *optname, zval *value, uint32_t arg)
+{
+	TidyOption opt = tidyGetOptionByName(doc, optname);
+	zend_long lval;
+
+	if (!opt) {
+		zend_argument_value_error(arg, "Unknown Tidy configuration option \"%s\"", optname);
+		return false;
+	}
+
+#if defined(HAVE_TIDYOPTGETCATEGORY)
+	if (tidyOptGetCategory(opt) == TidyInternalCategory) {
+#else
+	if (tidyOptIsReadOnly(opt)) {
+#endif
+		zend_argument_value_error(arg, "Attempting to set read-only option \"%s\"", optname);
+		return false;
+	}
+
+	TidyOptionType type = tidyOptGetType(opt);
+	if (type == TidyString) {
+		zend_string *tmp_str;
+		const zend_string *str = zval_get_tmp_string(value, &tmp_str);
+		const bool result = tidyOptSetValue(doc, tidyOptGetId(opt), ZSTR_VAL(str));
+		if (UNEXPECTED(!result)) {
+			zend_argument_type_error(arg, "option \"%s\" does not accept \"%s\" as a value", optname, ZSTR_VAL(str));
+		}
+		zend_tmp_string_release(tmp_str);
+		return result;
+	} else if (type == TidyInteger) {
+		lval = zval_get_long(value);
+		return tidyOptSetInt(doc, tidyOptGetId(opt), lval);
+	} else {
+		ZEND_ASSERT(type == TidyBoolean);
+		lval = zval_get_long(value);
+		return tidyOptSetBool(doc, tidyOptGetId(opt), lval);
+	}
+}
+
 static zend_result _php_tidy_apply_config_array(TidyDoc doc, const HashTable *ht_options, uint32_t arg)
 {
 	zval *opt_val;
 	zend_string *opt_name;
 
 	if (!HT_IS_PACKED(ht_options)) {
+		bool has_failures = false;
 		ZEND_HASH_MAP_FOREACH_STR_KEY_VAL(ht_options, opt_name, opt_val) {
 			if (opt_name == NULL) {
 				zend_argument_type_error(arg, "must be of type array with keys as string");
 				return FAILURE;
 			}
-			_php_tidy_set_tidy_opt(doc, ZSTR_VAL(opt_name), opt_val, arg);
+			has_failures = has_failures || !php_tidy_set_tidy_opt(doc, ZSTR_VAL(opt_name), opt_val, arg);
 		} ZEND_HASH_FOREACH_END();
-		return SUCCESS;
+		return has_failures ? FAILURE : SUCCESS;
 	} else {
 		zend_argument_type_error(arg, "must be of type array with keys as string");
 		return FAILURE;
 	}
 }
 
-static zend_result php_tidy_parse_string(PHPTidyObj *obj, const char *string, uint32_t len, const char *enc)
+static zend_result php_tidy_parse_string(PHPTidyObj *obj, const zend_string *string, const char *enc)
 {
 	TidyBuffer buf;
 
-	if(enc) {
+	ZEND_ASSERT(!ZEND_SIZE_T_UINT_OVFL(ZSTR_LEN(string)));
```

Comment:
do not know if an exception should be thrown instead or this. Early morning review so forgive me if I miss something obvious :)

Reply:
This was always checked before calling this function, just making sure it stays that way :)